### ISSUE TITLE
ArgoCD SSO/Token Issue

### ISSUE THREAD

**Issue by mattpoel** [OpenUnison/openunison-k8s#116]
**Describe the bug**

I've configured the OpenUnison SSO login according to the ArgoCD documentation, but somehow I keep having a login "loop" on the web interface. Everytime I press the login via openunison button, I see a glimpse of the logged in web page and will directly get forwarded again to login.

Login itself looks fine with the ArgoCD CLI:

```
# argocd login argo.demo.domain.com --sso
WARN[0001] Failed to invoke grpc call. Use flag --grpc-web in grpc calls. To avoid this warning message, use flag --grpc-web.
Opening browser for authentication
Performing authorization_code flow login: https://k8sou.demo.domain.com/auth/idp/k8sIdp/auth?access_type=offline&client_id=argocd&code_challenge=asdfasdf&code_challenge_method=S256&redirect_uri=http%3A%2F%2Flocalhost%3A8085%2Fauth%2Fcallback&response_type=code&scope=openid+profile+email+groups&state=yuuBWTNiDPQsXyOZtYygPKtQ
Authentication successful
' ' logged in successfully
Context 'argo.demo.domain.com' updated
```

Follow up command will either report an error or in this case let me know that I'm not logged in:

```
# argocd account get-user-info
Logged In: false
```

When checking the logs they mention a token verification failure with a tls certificate unknown authority error message, but certificates are signed with Let's Encrypt for OpenUnison and ArgoCD:

```
time="2024-06-19T20:08:28Z" level=warning msg="Failed to verify token: failed to verify token: token verification failed for all audiences: error for aud \"argocd\": Failed to query provider \"https://k8sou.demo.domain.com/auth/idp/k8sIdp\": Get \"https://k8sou.demo.domain.com/auth/idp/k8sIdp/.well-known/openid-configuration\": tls: failed to verify certificate: x509: certificate signed by unknown authority"
```

If I test the mentioned url in my browser or in a debug container, certificate is valid without any issue.

**To Reproduce**

Deploy current stable argocd version and configure OpenUnison SSO according to the documentation.

**Expected behavior**

Sign into ArgoCD with OpenUnison.

**Screenshots**

Screenshot from login "loop":

![login](https://github.com/argoproj/argo-cd/assets/15014728/29acd294-996c-44eb-9c1a-c8a8836ef1cf)

**Logs**

From argocd-server on login:

```
time="2024-06-20T10:28:31Z" level=info msg="Loading TLS configuration from secret argocd/argocd-server-tls"
time="2024-06-20T10:28:31Z" level=info msg="received unary call /cluster.SettingsService/Get" grpc.method=Get grpc.request.content= grpc.service=cluster.SettingsService grpc.start_time="2024-06-20T10:28:31Z" span.kind=server system=grpc
time="2024-06-20T10:28:31Z" level=info msg="Loading TLS configuration from secret argocd/argocd-server-tls"
time="2024-06-20T10:28:31Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=Get grpc.service=cluster.SettingsService grpc.start_time="2024-06-20T10:28:31Z" grpc.time_ms=4.241 span.kind=server system=grpc
```

From openunison-orchestra on login:

```
[2024-06-20 10:28:20,693][XNIO-1 task-9] INFO  AccessLog - [AzSuccess] - k8sidp - https://127.0.0.1:8443/auth/idp/k8sIdp/.well-known/openid-configuration - uid=Anonymous,o=Tremolo - NONE [127.0.0.1] - [f5fb0914c83f93b16e0977d7b5091818d55852b29]
[2024-06-20 10:28:20,696][XNIO-1 task-9] INFO  AccessLog - [AzSuccess] - k8sidp - https://127.0.0.1:8443/auth/idp/k8sIdp/.well-known/openid-configuration - uid=Anonymous,o=Tremolo - NONE [127.0.0.1] - [f7b36709e7550ee125526ab411bf3ec4250f89f06]
[2024-06-20 10:28:30,625][XNIO-1 task-9] INFO  AccessLog - [AzSuccess] - k8sidp - https://127.0.0.1:8443/auth/idp/k8sIdp/.well-known/openid-configuration - uid=Anonymous,o=Tremolo - NONE [127.0.0.1] - [f846c5b6b5b61d6c74beac88551c118623ee66bc8]
[2024-06-20 10:28:30,627][XNIO-1 task-12] INFO  AccessLog - [AzSuccess] - k8sidp - https://127.0.0.1:8443/auth/idp/k8sIdp/.well-known/openid-configuration - uid=Anonymous,o=Tremolo - NONE [127.0.0.1] - [f2a41025a037a52587bf9b36393c76ba6f9636bc2]
[2024-06-20 10:28:32,208][XNIO-1 task-12] INFO  AccessLog - [AzSuccess] - k8sidp - https://k8sou.demo.domain.com/auth/idp/k8sIdp/.well-known/openid-configuration - uid=Anonymous,o=Tremolo - NONE [10.0.23.71] - [f7ba0ead215ff92f883ee5026440b328395ced5c1]
[2024-06-20 10:28:33,601][XNIO-1 task-12] INFO  AccessLog - [AzSuccess] - k8sidp - https://k8sou.demo.domain.com/auth/idp/k8sidp/completefed - uid=x-54-xx-54-xx-54-xx-49-xx-55-xx-55-xx-50-xx-51-xx-51-xx-48-xfx-52-xdx-56-xbdx-57-xx-53-xx-50-xx-48-xx-52-xcx-56-xa,ou=shadow,o=Tremolo - NONE [10.0.23.71] - [f205af1c28915b02c9eb8b48fe22bc2fd2ea0f0ac]
[2024-06-20 10:28:33,677][XNIO-1 task-12] INFO  AccessLog - SRCH op=56 con=55 base='uid=x-54-xx-54-xx-54-xx-49-xx-55-xx-55-xx-50-xx-51-xx-51-xx-48-xfx-52-xdx-56-xbdx-57-xx-53-xx-50-xx-48-xx-52-xcx-56-xa,ou=shadow,o=Tremolo' filter='(objectClass=*)' scope='0' attribs=''
[2024-06-20 10:28:33,697][XNIO-1 task-12] INFO  AccessLog - RESULT op=56 con=55 result=0 time=20
[2024-06-20 10:28:33,697][XNIO-1 task-12] INFO  AccessLog - SRCH-RESULT op=56 con=55 entries=1 time=20
[2024-06-20 10:28:33,700][XNIO-1 task-12] INFO  AccessLog - SRCH op=57 con=56 base='uid=x-54-xx-54-xx-54-xx-49-xx-55-xx-55-xx-50-xx-51-xx-51-xx-48-xfx-52-xdx-56-xbdx-57-xx-53-xx-50-xx-48-xx-52-xcx-56-xa,ou=shadow,o=Tremolo' filter='(objectClass=*)' scope='0' attribs=''
[2024-06-20 10:28:33,709][XNIO-1 task-12] INFO  AccessLog - RESULT op=57 con=56 result=0 time=9
[2024-06-20 10:28:33,710][XNIO-1 task-12] INFO  AccessLog - SRCH-RESULT op=57 con=56 entries=1 time=10
[2024-06-20 10:28:33,713][XNIO-1 task-12] WARN  K8sSessionStore - Could not find trust argocd, not storing expired tokens
[2024-06-20 10:28:33,738][XNIO-1 task-12] INFO  OpenShiftTarget - DR Queues Size : 0
[2024-06-20 10:28:33,740][XNIO-1 task-12] INFO  AccessLog - [AzSuccess] - k8sidp - https://k8sou.demo.domain.com/auth/idp/k8sIdp/token - uid=x-54-xx-54-xx-54-xx-49-xx-55-xx-55-xx-50-xx-51-xx-51-xx-48-xfx-52-xdx-56-xbdx-57-xx-53-xx-50-xx-48-xx-52-xcx-56-xa,ou=shadow,o=Tremolo - NONE [10.0.23.71] - [f591d487ef374a524f3ed5c1c313f43197318dd65]
[2024-06-20 10:28:36,590][XNIO-1 task-12] INFO  AccessLog - [AzSuccess] - scale-session-check - https://k8sou.demo.domain.com/scale/sessioncheck - uid=Anonymous,o=Tremolo - NONE [10.0.23.71] - [ff05238c1fefb239ca978e1ca38372590f57e9bd8]
[2024-06-20 10:28:40,670][XNIO-1 task-12] INFO  AccessLog - [AzSuccess] - k8sidp - https://127.0.0.1:8443/auth/idp/k8sIdp/.well-known/openid-configuration - uid=Anonymous,o=Tremolo - NONE [127.0.0.1] - [f227fe5771a39f2fd79091ea881b28baf065275e8]
[2024-06-20 10:28:40,670][XNIO-1 task-9] INFO  AccessLog - [AzSuccess] - k8sidp - https://127.0.0.1:8443/auth/idp/k8sIdp/.well-known/openid-configuration - uid=Anonymous,o=Tremolo - NONE [127.0.0.1] - [f601d098b8a981b23483680ab686fa87e9dfe2293]
```

From argocd-server on an actual command executed by argocd:

```
time="2024-06-20T10:29:02Z" level=info msg="Loading TLS configuration from secret argocd/argocd-server-tls"
time="2024-06-20T10:29:02Z" level=info msg="Initializing OIDC provider (issuer: https://k8sou.demo.domain.com/auth/idp/k8sIdp)"
time="2024-06-20T10:29:02Z" level=warning msg="Failed to verify token: failed to verify token: token verification failed for all audiences: error for aud \"argocd\": Failed to query provider \"https://k8sou.demo.domain.com/auth/idp/k8sIdp\": Get \"https://k8sou.demo.domain.com/auth/idp/k8sIdp/.well-known/openid-configuration\": tls: failed to verify certificate: x509: certificate signed by unknown authority"
time="2024-06-20T10:29:02Z" level=info msg="Loading TLS configuration from secret argocd/argocd-server-tls"
time="2024-06-20T10:29:02Z" level=info msg="received unary call /session.SessionService/GetUserInfo" grpc.method=GetUserInfo grpc.request.content= grpc.service=session.SessionService grpc.start_time="2024-06-20T10:29:02Z" span.kind=server system=grpc
time="2024-06-20T10:29:02Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=GetUserInfo grpc.service=session.SessionService grpc.start_time="2024-06-20T10:29:02Z" grpc.time_ms=24.488 span.kind=server system=grpc
```

From openunison-orchestra on an actual command executed by argocd:

```
[2024-06-20 10:29:01,745][XNIO-1 task-12] INFO  AccessLog - [AzSuccess] - scale-session-check - https://k8sou.demo.domain.com/scale/sessioncheck - uid=Anonymous,o=Tremolo - NONE [10.0.23.71] - [fda999128b46c8602b7e28bb6251ca8b317d85192]
[2024-06-20 10:29:10,610][XNIO-1 task-12] INFO  AccessLog - [AzSuccess] - k8sidp - https://127.0.0.1:8443/auth/idp/k8sIdp/.well-known/openid-configuration - uid=Anonymous,o=Tremolo - NONE [127.0.0.1] - [f08dfc7fc5fa7d3d9dc5426ca05bdb0e396596465]
[2024-06-20 10:29:10,674][XNIO-1 task-12] INFO  AccessLog - [AzSuccess] - k8sidp - https://127.0.0.1:8443/auth/idp/k8sIdp/.well-known/openid-configuration - uid=Anonymous,o=Tremolo - NONE [127.0.0.1] - [f954373c7498201b0c6a19ef033d3d741ec529136]
[2024-06-20 10:29:10,686][Thread-24] INFO  K8sWatcher - Resource 4770088 already processed, skipping
[2024-06-20 10:29:11,705][Thread-24] INFO  K8sWatcher - watching https://kubernetes.default.svc/apis/openunison.tremolo.io/v1/namespaces/openunison/trusts?watch=true&timeoutSecond=25&allowWatchBookmarks=true&resourceVersion=4770088
[2024-06-20 10:29:14,735][Thread-21] WARN  OpenShiftTarget - Unexpected result calling 'https://kubernetes.default.svc/apis/openunison.tremolo.io/v2/namespaces/openunison/oidc-sessions/x0f234c5a-fb46-455a-ac55-2d218552a0fcx' - 404 / {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"oidc-sessions.openunison.tremolo.io \"x0f234c5a-fb46-455a-ac55-2d218552a0fcx\" not found","reason":"NotFound","details":{"name":"x0f234c5a-fb46-455a-ac55-2d218552a0fcx","group":"openunison.tremolo.io","kind":"oidc-sessions"},"code":404}

[2024-06-20 10:29:14,755][Thread-21] WARN  OpenShiftTarget - Unexpected result calling 'https://kubernetes.default.svc/apis/openunison.tremolo.io/v2/namespaces/openunison/oidc-sessions/x03a59a74-940a-4288-bb2b-9706d9f2f581x' - 404 / {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"oidc-sessions.openunison.tremolo.io \"x03a59a74-940a-4288-bb2b-9706d9f2f581x\" not found","reason":"NotFound","details":{"name":"x03a59a74-940a-4288-bb2b-9706d9f2f581x","group":"openunison.tremolo.io","kind":"oidc-sessions"},"code":404}

[2024-06-20 10:29:14,771][Thread-21] WARN  K8sSessionStore - Could not find trust argocd, not loading expired tokens
[2024-06-20 10:29:14,802][Thread-21] WARN  OpenShiftTarget - Unexpected result calling 'https://kubernetes.default.svc/apis/openunison.tremolo.io/v2/namespaces/openunison/oidc-sessions/xd5bddca2-bc7e-4cf9-83a2-bbd0510c4ac3x' - 404 / {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"oidc-sessions.openunison.tremolo.io \"xd5bddca2-bc7e-4cf9-83a2-bbd0510c4ac3x\" not found","reason":"NotFound","details":{"name":"xd5bddca2-bc7e-4cf9-83a2-bbd0510c4ac3x","group":"openunison.tremolo.io","kind":"oidc-sessions"},"code":404}

[2024-06-20 10:29:14,818][Thread-21] WARN  K8sSessionStore - Could not find trust argocd, not loading expired tokens
[2024-06-20 10:29:14,834][Thread-21] WARN  OpenShiftTarget - Unexpected result calling 'https://kubernetes.default.svc/apis/openunison.tremolo.io/v2/namespaces/openunison/oidc-sessions/x59f840e8-b2f3-46d8-a549-302213298ecex' - 404 / {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"oidc-sessions.openunison.tremolo.io \"x59f840e8-b2f3-46d8-a549-302213298ecex\" not found","reason":"NotFound","details":{"name":"x59f840e8-b2f3-46d8-a549-302213298ecex","group":"openunison.tremolo.io","kind":"oidc-sessions"},"code":404}

[2024-06-20 10:29:14,874][Thread-21] WARN  OpenShiftTarget - Unexpected result calling 'https://kubernetes.default.svc/apis/openunison.tremolo.io/v2/namespaces/openunison/oidc-sessions/x57078fae-09c1-4854-af9f-1020a73b73eax' - 404 / {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"oidc-sessions.openunison.tremolo.io \"x57078fae-09c1-4854-af9f-1020a73b73eax\" not found","reason":"NotFound","details":{"name":"x57078fae-09c1-4854-af9f-1020a73b73eax","group":"openunison.tremolo.io","kind":"oidc-sessions"},"code":404}

[2024-06-20 10:29:14,901][Thread-21] WARN  OpenShiftTarget - Unexpected result calling 'https://kubernetes.default.svc/apis/openunison.tremolo.io/v2/namespaces/openunison/oidc-sessions/xf4e68bf7-563f-4653-9db3-facfdf9732e2x' - 404 / {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"oidc-sessions.openunison.tremolo.io \"xf4e68bf7-563f-4653-9db3-facfdf9732e2x\" not found","reason":"NotFound","details":{"name":"xf4e68bf7-563f-4653-9db3-facfdf9732e2x","group":"openunison.tremolo.io","kind":"oidc-sessions"},"code":404}

[2024-06-20 10:29:14,926][Thread-21] WARN  OpenShiftTarget - Unexpected result calling 'https://kubernetes.default.svc/apis/openunison.tremolo.io/v2/namespaces/openunison/oidc-sessions/x92896cc7-4430-4dae-ab3c-28bf044b0b37x' - 404 / {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"oidc-sessions.openunison.tremolo.io \"x92896cc7-4430-4dae-ab3c-28bf044b0b37x\" not found","reason":"NotFound","details":{"name":"x92896cc7-4430-4dae-ab3c-28bf044b0b37x","group":"openunison.tremolo.io","kind":"oidc-sessions"},"code":404}

[2024-06-20 10:29:14,946][Thread-21] WARN  SessionManagerImpl - Clearing 14 sessions
```

Thanks a lot for your help in advance!

**Comment by mattpoel** [comment]
They are on the same EKS cluster, but the argocd-server and openunison-operator are on different nodes. It's a managed node group, therefore time synchronization shouldn't be an issue (hopefully).

Double checked with simultaneous date execution:

```
openunison@openunison-orchestra-648555b6b-nt892:/$ date
Thu Jun 20 13:34:08 UTC 2024

argocd@argocd-server-89799d7ff-6tcdt:~$ date
Thu Jun 20 13:34:08 UTC 2024
```

**Comment by mlbiam** [comment]
> time synchronization shouldn't be an issue

yep, have your tried in an incognito window?  Also, try destroying all the containers involved (openunison and argocd) and let them come back.

**Comment by mattpoel** [comment]
@mlbiam I'm pretty sure I did, but thought well another time can't do any harm, but let's also remove the OIDC configuration in ArgoCD :wink:

This is how I got it running:

* Removing the OIDC configuration from the `argocd-cm` ConfigMap.
* Deleting the `argocd-server` pod (to get a new one running).
* Re-adding the OIDC configuration to the `argocd-cm` ConfigMap.
* Deleting the `argocd-server` pod again to get a new one up and running.
* Reloading my ArgoCD browser tab.

:boom: I was in :metal:

I guess the restart/recreation of the `argocd-server` pod fixed this.

Thanks for your support!
